cmd/nerdctl/cp.go

```diff
@@ -0,0 +1,229 @@
+
+	var cpCommand = &cobra.Command{
+		Use:   "cp [OPTIONS] CONTAINER:SRC_PATH DEST_PATH|-",
+		Args:  cobra.MinimumNArgs(2),
+		Short: shortHelp,
+		Long:  longHelp,
+		RunE:  cpAction,
+		//ValidArgsFunction: cpShellComplete,
+		SilenceUsage:  true,
+		SilenceErrors: true,
+	}
+
+	cpCommand.Flags().BoolP("follow-link", "L", false, "Always follow symbol link in SRC_PATH")
+	return cpCommand
+}
+
+func cpAction(cmd *cobra.Command, args []string) error {
+	srcSpec, err := splitCpArg(args[0])
+	if err != nil {
+		return err
+	}
+
+	destSpec, err := splitCpArg(args[0])
+	if err != nil {
+		return err
+	}
+
+	flagL, err := cmd.Flags().GetBool("follow-link")
+	if err != nil {
+		return err
+	}
+
+	copyConfig := cpConfig{
+		followLink: flagL,
+		sourcePath: srcSpec.Path,
+		destPath:   destSpec.Path,
+	}
+
+	if len(srcSpec.Container) != 0 && len(destSpec.Container) != 0 {
+		return fmt.Errorf("one of src or dest must be a local file specification")
+	}
+	if len(srcSpec.Path) == 0 || len(destSpec.Path) == 0 {
+		return errors.New("filepath can not be empty")
+	}
+
+	_, ctx, cancel, err := newClient(cmd)
+	if err != nil {
+		return err
+	}
+	defer cancel()
+
+	if len(srcSpec.Container) != 0 {
+		copyConfig.container = srcSpec.Container
+		return copyFromContainer(ctx, cmd, copyConfig)
+	}
+	if len(destSpec.Container) != 0 {
+		copyConfig.container = destSpec.Container
+		return nil
+	}
+	return fmt.Errorf("one of src or dest must be a remote file specification")
+
+}
+
+func splitCpArg(arg string) (fileSpec, error) {
+	i := strings.Index(arg, ":")
+
+	// filespec starting with a semicolon is invalid
+	if i == 0 {
+		return fileSpec{}, errFileSpecDoesntMatchFormat
+	}
+
+	if system.IsAbs(arg) {
+		// Explicit local absolute path, e.g., `C:\foo` or `/foo`.
+		return fileSpec{
+			Path: arg,
+		}, nil
+	}
+
+	parts := strings.SplitN(arg, ":", 2)
+
+	if len(parts) == 1 || strings.HasPrefix(parts[0], ".") {
+		// Either there's no `:` in the arg
+		// OR it's an explicit local relative path like `./file:name.txt`.
+		return fileSpec{
+			Path: arg,
+		}, nil
+	}
+
+	return fileSpec{
+		Container: parts[0],
+		Path:      parts[1],
+	}, nil
+}
+
+type fileSpec struct {
+	Container string
+	Path      string
+}
+
+func copyFromContainer(ctx context.Context, cmd *cobra.Command, copyConfig cpConfig) (err error) {
+	dstPath := copyConfig.destPath
+	srcPath := copyConfig.sourcePath
+
+	if dstPath != "-" {
+		if dstPath, err = filepath.Abs(dstPath); err != nil {
+			return err
+		}
+	}
+
+	dir := filepath.Dir(filepath.Clean(dstPath))
+	if dir != "" && dir != "." {
+		if _, err := os.Stat(dir); os.IsNotExist(err) {
+			return fmt.Errorf("invalid output path: directory %q does not exist", dir)
+		}
+	}
+
+	reader := newTarPipe(src, o)
+
+	tarReader := tar.NewReader(reader)
```

Comment:
Can we just exec `cp -a SRC /proc/PID/root/DST` without using tar?

Reply:
`Cp` is safer than `tar `  🤔 !

Reply:
@AkihiroSuda we have `rsync` too, but `tar`  is still faster and more secure than `cp` and `rsync`. WDYT ?

Reply:
I'm also thinking about the day when we will support remote containers (https://github.com/containerd/nerdctl/issues/473) .  A bulk filesystem copy with `cp` across a network will be very slow.

Reply:
I think we should keep using tar and `notifying` user about possible dangers